core/src/main/java/com/datastax/oss/driver/internal/core/os/AbstractNativeImpl.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.os;
+
+import java.util.function.Supplier;
+
+public abstract class AbstractNativeImpl implements NativeImpl {
+
+  Supplier<IllegalStateException> gettimeofdaySupplier =
```

Comment:
AbstractNativeImpl was a fairly late addition to the process, and it's entire reason for existence is to ensure a common exception experience across all concrete instances of NativeImpl.  I have no problem removing it if we can find an alternate way to address it's behaviours.

I also very much like the idea of removing the "is available" methods and just having getX() return an optional type, but note the consequence here.  We can no longer distinguish between a case in which the concrete type couldn't complete it's setup (i.e. when JNR wasn't present or failed to initialize) vs. when a specific function call failed.  In each case we'll return an Optional.empty() and Native will have to do... something.  We could chuck the IllegalStateException in that case, but if we do that the text will have to be tweaked (since it doesn't necessarily indicate that the native call isn't available).

A reasonable (to me) answer here would be to log the specific error in JnrNativeImpl and then throw the IllegalStateException at the Native level saying something like "native call did not succeed, consult logs for specific errors".  An Either type would also be great here but let's not hope for wild and crazy things. :)

Reply:
I took a stab at what this might look like in https://github.com/datastax/java-driver/pull/1425.  Take a look and see if this is preferred... I can definitely see some benefit there.

Reply:
This has now been merged into this feature branch